third_party/move/move-vm/runtime/src/loader/mod.rs

```diff
@@ -307,16 +315,27 @@ impl Loader {
 
         let mut scripts = self.scripts.write();
         let main = match scripts.get(&hash_value) {
-            Some(cached) => cached,
+            Some(ScriptCacheEntry::Verified(script)) => script.entry_point(),
+            Some(ScriptCacheEntry::Deserialized(compiled_script)) => {
+                self.verify_script(compiled_script.as_ref(), data_store, module_store)?;
+                let script = Script::new(
+                    Arc::clone(compiled_script),
+                    &hash_value,
+                    module_store,
+                    &self.name_cache,
+                )?;
+                scripts.insert_verified_script(hash_value, script)
+            },
             None => {
-                let ver_script = self.deserialize_and_verify_script(
-                    script_blob,
-                    hash_value,
-                    data_store,
+                let compiled_script = self.deserialize_script(script_blob)?;
+                self.verify_script(&compiled_script, data_store, module_store)?;
```

Comment:
hm, the comment on verify_script (previously on deserialize_and_verify) said it should not happen under a lock, but as far as I can see, it happens under scripts write lock? should we give up and re-acquire?